storage/writer.go

```diff
@@ -84,30 +94,301 @@ type Writer struct {
+
+			o, off, finalized, err := w.upload(buf, recvd, offset, done, upid)
+			if err != nil {
+				w.error(err)
+				pr.CloseWithError(err)
+				return
+			}
+			offset = off
+
+			// When we are done reading data and the chunk has been finalized,
+			// we are done.
+			if done && finalized {
+				// Build Object from server's response and report progress.
+				w.obj = newObjectFromProto(o)
+				w.progress(o.GetSize())
+				return
+			}
+
+			// Query the progress then upload another chunk.
+			if err := w.queryProgress(upid); err != nil {
+				w.error(err)
+				pr.CloseWithError(err)
+				return
+			}
+		}
+	}()
+
+	return nil
+}
+
+// startResumableUpload initializes a Resumable Upload and returns the upload ID.
+//
+// This is an experimental API and not intended for public use.
+func (w *Writer) startResumableUpload() (string, error) {
+	var common *storagepb.CommonRequestParams
+	if w.o.userProject != "" {
+		common = &storagepb.CommonRequestParams{UserProject: w.o.userProject}
 	}
-	if w.ChunkSize < 0 {
-		return errors.New("storage: Writer.ChunkSize must be non-negative")
+	upres, err := w.o.c.gc.StartResumableWrite(w.ctx, &storagepb.StartResumableWriteRequest{
+		WriteObjectSpec: &storagepb.WriteObjectSpec{
+			Resource: w.ObjectAttrs.toProtoObject(w.o.bucket),
+		},
+		CommonRequestParams: common,
+	})
+
+	return upres.GetUploadId(), err
+}
+
+// queryProgress is a helper that queries the status of the resumable upload
+// associated with the given upload ID. Progress is reported with the committed
+// size from the write status.
+//
+// This is an experimental API and not intended for public use.
+func (w *Writer) queryProgress(upid string) error {
+	q, err := w.o.c.gc.QueryWriteStatus(w.ctx, &storagepb.QueryWriteStatusRequest{UploadId: upid})
+
+	// q.GetCommittedSize() will return 0 if q is nil, and progress() will ignore 0 progress.
+	w.progress(q.GetCommittedSize())
+	return err
+}
+
+// upload opens a Write stream and uploads the buffer at the given offset (if
+// uploading a chunk for a resumable upload), and will mark the write as
+// finished if we are done receiving data from the user. The resulting write
+// offset after uploading the buffer is returned, as well as a boolean
+// indicating if the Object has been finalized. If it has been finalized, the
+// final Object will be returned as well.
+//
+// This is an experimental API and not intended for public use.
+func (w *Writer) upload(buf []byte, recvd int, offset int64, done bool, upid string) (*storagepb.Object, int64, bool, error) {
+	var err error
+
+	sent := 0
+	first := true
+	finished := false
+	limit := maxPerMessageWriteSize
+	for recvd-sent > 0 {
+		// This indicates that this is the last message and the remaining
+		// data fits in one message.
+		if sent+limit >= recvd {
+			limit = recvd - sent
+			finished = true
+
+			// Do not indicate finished for Resumable Uploads until we have
+			// received all data for writing from the user.
+			if !done && upid != "" {
```

Comment:
I think it makes more sense to only set finished to true if `done` right?

Reply:
Well `done` in this case is if we are done reading from the pipe i.e. we have received an EOF. I changed the name to `doneReading` because `done` is an admittedly ambiguous name. If we just filled up our buffer, but the writer hasn't closed their end, this is a resumable upload and we cannot say `finished = true`.

Reply:
Fixed this together offline.